src/ripple/ledger/impl/ApplyStateTable.cpp

```diff
@@ -0,0 +1,589 @@
+    {
+    case Action::erase:
+        return false;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    }
+    if (! k.check(*sle))
+        return false;
+    return true;
+}
+
+auto
+ApplyStateTable::succ (ReadView const& base,
+    key_type const& key, boost::optional<
+        key_type> const& last) const ->
+            boost::optional<key_type>
+{
+    boost::optional<key_type> next = key;
+    items_t::const_iterator iter;
+    // Find base successor that is
+    // not also deleted in our list
+    do
+    {
+        next = base.succ(*next, last);
+        if (! next)
+            break;
+        iter = items_.find(*next);
+    }
+    while (iter != items_.end() &&
+        iter->second.first == Action::erase);
+    // Find non-deleted successor in our list
+    for (iter = items_.upper_bound(key);
+        iter != items_.end (); ++iter)
+    {
+        if (iter->second.first != Action::erase)
+        {
+            // Found both, return the lower key
+            if (! next || next > iter->first)
+                next = iter->first;
+            break;
+        }
+    }
+    // Nothing in our list, return
+    // what we got from the parent.
+    if (last && next >= last)
+        return boost::none;
+    return next;
+}
+
+std::shared_ptr<SLE const>
+ApplyStateTable::read (ReadView const& base,
+    Keylet const& k) const
+{
+    auto const iter = items_.find(k.key);
+    if (iter == items_.end())
+        return base.read(k);
+    auto const& item = iter->second;
+    auto const& sle = item.second;
+    switch (item.first)
+    {
+    case Action::erase:
+        return nullptr;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    };
+    if (! k.check(*sle))
+        return nullptr;
+    return sle;
+}
+
+std::shared_ptr<SLE>
+ApplyStateTable::peek (ReadView const& base,
+    Keylet const& k)
+{
+    auto iter = items_.lower_bound(k.key);
+    if (iter == items_.end() ||
+        iter->first != k.key)
+    {
+        auto const sle = base.read(k);
+        if (! sle)
+            return nullptr;
+        // Make our own copy
+        using namespace std;
+        iter = items_.emplace_hint (iter,
+            piecewise_construct,
+                forward_as_tuple(sle->key()),
+                    forward_as_tuple(Action::cache,
+                        make_shared<SLE>(*sle)));
+        return iter->second.second;
+    }
+    auto const& item = iter->second;
+    auto const& sle = item.second;
+    switch (item.first)
+    {
+    case Action::erase:
+        return nullptr;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    };
+    if (! k.check(*sle))
+        return nullptr;
+    return sle;
+}
+
+void
+ApplyStateTable::erase(
+    ReadView const& base,
+        std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.find(sle->key());
+    if (iter == items_.end())
```

Comment:
This analysis is simply wrong. The `erase` function terminates the lifetime of an object and specifies its final state. If the object doesn't exist, then it did neither of these things. This is part of the digital chain of custody of a cryptographically-secured object that reflects financial information. Code breaching core rules for how those objects are treated should not be permitted to continue and commit its changes to a public, immutable structure.